source/common/upstream/cluster_manager_impl.cc

```diff
@@ -206,8 +207,49 @@ ClusterManagerImpl::ClusterManagerImpl(const envoy::api::v2::Bootstrap& bootstra
         bootstrap.cluster_manager().upstream_bind_config().source_address());
   }
 
+  // set of statically loaded cluster names used for further config checks
+  std::unordered_set<std::string> cluster_names;
+
   for (const auto& cluster : bootstrap.static_resources().clusters()) {
     loadCluster(cluster, false);
+    cluster_names.insert(cluster.name());
+  }
+
+  // All the static clusters have been loaded. At this point we can check for the
+  // existence of the v1 sds backing cluster, and the ads backing cluster.
+  // TODO(htuch): Add support for multiple clusters, #1170.
+  if (bootstrap.dynamic_resources().deprecated_v1().has_sds_config()) {
+    const auto& sds_config = bootstrap.dynamic_resources().deprecated_v1().sds_config();
+    switch (sds_config.config_source_specifier_case()) {
```

Comment:
This code is similar to that added in `SubscriptionFactory`, operating on a `ApiConfigSource`, would it make sense to refactor?